packages/abi-coder/src/function-fragment.ts

```diff
@@ -18,27 +18,29 @@ import type {
   JsonAbiFunction,
   JsonAbiFunctionAttribute,
 } from './json-abi';
+import { ResolvedAbiType } from './resolved-abi-type';
 import type { Uint8ArrayWithDynamicData } from './utilities';
 import { isPointerType, unpackDynamicData, findOrThrow } from './utilities';
 
 const logger = new Logger(versions.FUELS);
 
 export class FunctionFragment<
-  TAbi extends JsonAbi = JsonAbi,
-  FnName extends TAbi['functions'][number]['name'] = string
+  Input extends object | never = never,
```

Comment:
Yes, now all inputs are objects by necessity, because TS doesn't allow giving generic names to parameterized tuples: `[a: number, b: string]` -> `<myGenericParam>:<myGenericParamType>[]` (pseudo code, not allowed). Due to this, I had to change the input into an object with keys that match parameter names:
```rs
// sway  
struct MyStruct {
prop1: bool;
prop2: b512;
}
fn my_fn(a: u32, b: b256, c: MyStruct);
```
```ts
// input for my_fn
type InferredInputForMyFn = {
 a: number;
 b: string;
 c: {
  prop1: boolean;
  prop2: string;
 }
}
```

Reply:
I can change the type inference to be arrays as it was previously, but we would lose info on names of sway function arguments, because TS can't generate tuples with named elements (e.g. `[arg1: number, arg2: string]` - this is valid TS syntax, but it can't be generated dynamically). 

However, your idea on allowing both approaches is great! We can do a hybrid approach with [function overloads](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads). This wouldn't break **any** existing application, but we can make it so that the first approach suggested by the IDE is now the "object for all inputs" instead of "spread array inputs".

Reply:
> However, your idea on allowing both approaches is great! We can do a hybrid approach with [function overloads](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads). This wouldn't break **any** existing application, but we can make it so that the first approach suggested by the IDE is now the "object for all inputs" instead of "spread array inputs".

I've updated the code to now have the overload and it works! The `Input` type that @camsjams  commented on is still an object (or, more specifically `Record<string, unknown>`), but the actual `encodeArguments` method can now accept both a tuple as it was accepting pre-type-inference, as well as a typesafe object. The typesafe tuple doesn't have named elements, but that's okay because I've ordered the overloads in a way that the IDE first suggests the object as an input, where the parameter names are keys of the object. The overload is specifically there to not introduce breaking changes - and it does its job as far as that is concerned.